gio/src/list_model.rs

```diff
@@ -19,15 +31,149 @@ impl<T: IsA<ListModel>> ListModelExtManual for T {
         }
         res
     }
+    fn iter<LT: IsA<glib::Object>>(&self) -> Result<ListModelIter<LT>, &Self>
+    where
+        Self: Sized + Clone,
+    {
+        if !self.item_type().is_a(LT::static_type()) {
+            return Err(self);
+        }
+
+        let len = self.n_items();
+        let changed = Rc::new(Cell::new(false));
+
+        let changed_clone = changed.clone();
+        let signal_id = Cell::new(Some(self.connect_items_changed(move |_, pos, _, _| {
+            changed_clone.replace(pos < len);
+        })));
+
+        Ok(ListModelIter {
+            ty: Default::default(),
+            i: 0,
+            reverse_pos: len,
+            model: self.clone().upcast(),
+            changed,
+            signal_id,
+        })
+    }
 }
 
-impl std::iter::IntoIterator for ListModel {
-    type Item = glib::Object;
-    type IntoIter = std::vec::IntoIter<glib::Object>;
+#[derive(thiserror::Error, Debug, PartialEq)]
+#[error("the list model was mutated during iteration")]
+pub struct ListModelMutatedDuringIter;
+
+// rustdoc-stripper-ignore-next
+/// Iterator of `ListModel`'s items.
+/// This iterator will always give `n = initial_model.n_items()` items, even if the `ListModel`
+/// is mutated during iteration.
+/// If the internal `ListModel` gets mutated, the iterator
+/// will return `Some(Err(...))` for the remaining items.
+/// Mutations to the `ListModel` in position >= `initial_model.n_items()` are allowed.
+pub struct ListModelIter<T: IsA<glib::Object>> {
+    ty: PhantomData<T>,
+    i: u32,
+    // it's > i when valid
+    reverse_pos: u32,
+    model: ListModel,
+    changed: Rc<Cell<bool>>,
+    signal_id: Cell<Option<SignalHandlerId>>,
+}
+impl<T: IsA<glib::Object>> Iterator for ListModelIter<T> {
+    type Item = Result<T, ListModelMutatedDuringIter>;
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let n = (self.reverse_pos - self.i) as usize;
+        (n as usize, Some(n))
+    }
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.i >= self.reverse_pos {
```

Comment:
They both work equally well. In this case I read `==` as "If both are equal, return early. The code below shouldn't be executed".
I prefer `>=` because it also says "The code below should be executed ONLY if `self.i < self.reverse_pos`".

But I see why you may like `==`: "`self.i` can't never be greater than `self.reverse_pos`, so why check?"

Whatever... :)